Theme options

redundans opened the pull request:
This Pull Request is an migration from saving items as featured content post types to saving them in them_mod option.

lakrisgubben commented:
Found a bug where if you have the same post in more than one area, they get the same id, which means that things such as click handlers will only affect to the first on in the dom. So if you go into the second featured area and try to open/edit a post that exists in the first area, the one in the first area will open, but not in the second. Maybe this affects other things as well, not sure? Having non-unique id's is forbidden anyways, better to target by something else I guess. :)

redundans commented:
> Found a bug where if you have the same post in more than one area, they get the same id, which means that things such as click handlers will only affect to the first on in the dom. So if you go into the second featured area and try to open/edit a post that exists in the first area, the one in the first area will open, but not in the second. Maybe this affects other things as well, not sure? Having non-unique id's is forbidden anyways, better to target by something else I guess. :)

I replaced all ID-queries with queries looking for data attributes in the featured area container. I hope that it will solve all the problems. Also added a ID for items element based on *featured_area* + "_item_" + *post_id*.

lakrisgubben commented on assets/js/customizer.js:
```diff
@@ -93,9 +93,7 @@
-							const parentItemOl = areaContainer.querySelector(
-								"#item_" + this.parent + " ol"
-							);
+							const parentItemOl = areaContainer.querySelector('[data-id="' + this.id + '"] ol');
```
pretty sure this should use this.parent not this.id @redundans

lakrisgubben replied:
fixed

lakrisgubben commented:
I think this is feature complete for a V1. The todo's that are left are just nice to haves to make the interface better for users. That is of course if we can't find any other bugs. :) But I cannot do that. Maybe test this with some data from Arbetaren that uses images, titles and excerpts @redundans?

lakrisgubben commented on assets/js/customizer.js:
```diff
@@ -7,108 +7,124 @@
-					this.element_id = "item_" + post.ID;
+					this.parent = parent;
+					this.element = null;
+					this.element_id = control.id + "_item_" + post.id;
```
I think we should use this.id instead of post.id here, just to be consistent. What do you say @redundans

lakrisgubben commented on assets/js/customizer.js:
```diff
@@ -7,108 +7,124 @@
-				constructor(post) {
-					this.key = post.ID;
+				constructor(post, parent) {
+					this.id = post.id;
```
To make this back-compat, i.e. work with theme settings from the current version of the plugin, we could replace this line with something like:

`if ( post.hasOwnProperty('id') ) {
						this.id = post.id;
					} else if ( post.hasOwnProperty('ID') ) {
						this.id = post.ID;
					} else {
						this.id = null;
					}`

lakrisgubben replied:
maybe this should check for original_post_id instead

lakrisgubben replied:
`if ( post.hasOwnProperty('original_post_id') ) {
						this.id = post.original_post_id;
						post.id = this.id;
					} else if ( post.hasOwnProperty('id') ) {
						this.id = post.id;
					} else {
						this.id = null;
					}`

lakrisgubben commented on assets/js/customizer.js:
```diff
@@ -7,108 +7,124 @@
+						this.element.id = this.element_id;
+						this.element.classList.add(this.postData.post_status);
+						for(var attribute in this.postData ) {
+							this.element.setAttribute('data-' + attribute, this.postData[attribute]);
```
Another small back-compat here would be to skip post_content if it exists, it's often very large and make the dom superslow when added as a data-attribute.

`if ( 'post_content' !== attribute ) {
								this.element.setAttribute('data-' + attribute, this.postData[attribute]);
							}`
